lambda/value_rec_compiler.ml

```diff
@@ -0,0 +1,787 @@
+     and h_lifted = fun x ->
+       incr h_context.v;
+       h_lifted (bar x)
+   ]}
+   Note that the closure environment computed from the context is passed as a
+   mutually recursive definition, that is, a free variable, and not as an
+   additional function parameter (which is customary for closure conversion).
+
+   Finally, when the tail expression is a variable, we perform an eta-expansion
+   to get a syntactic function, that we can then close and lift:
+   {[
+     let rec i_context =
+       let aux x = i_lifted (baz x) in
+       { aux }
+     and i_lifted = fun x -> i_context.aux x
+   ]}
+*)
+
+type lifted_function =
+  { lfun : Lambda.lfunction;
+    free_vars_block_size : int;
+  }
+
+type 'a split_result =
+  | Unreachable
+  | Reachable of lifted_function * 'a
+
+let ( let+ ) res f =
+  match res with
+  | Unreachable -> Unreachable
+  | Reachable (func, lam) -> Reachable (func, f lam)
+
+(* The closure blocks are immutable.
+   (Note: It is usually safe to declare immutable blocks as mutable,
+   but in this case the blocks might be empty and declaring them as Mutable
+   would cause errors later.) *)
+let lifted_block_mut : Asttypes.mutable_flag = Immutable
+
+let no_loc = Debuginfo.Scoped_location.Loc_unknown
+
+let rec split_static_function block_var local_idents lam :
+  Lambda.lambda split_result =
+  match lam with
+  | Lvar v ->
+    (* Eta-expand *)
+    (* Note: knowing the arity might let us generate slightly better code *)
+    let param = Ident.create_local "let_rec_param" in
+    let ap_func =
+      Lprim (Pfield (0, Pointer, lifted_block_mut), [Lvar block_var], no_loc)
+    in
+    let body =
+      Lapply {
+        ap_func;
+        ap_args = [Lvar param];
+        ap_loc = no_loc;
+        ap_tailcall = Default_tailcall;
+        ap_inlined = Default_inline;
+        ap_specialised = Default_specialise;
+      }
+    in
+    let wrapper =
+      lfunction'
+        ~kind:Curried
+        ~params:[param, Pgenval]
+        ~return:Pgenval
+        ~body
+        ~attr:default_stub_attribute
+        ~loc:no_loc
+    in
+    let lifted = { lfun = wrapper; free_vars_block_size = 1 } in
+    Reachable (lifted,
+               Lprim (Pmakeblock (0, lifted_block_mut, None), [Lvar v], no_loc))
+  | Lfunction lfun ->
+    let free_vars = Lambda.free_variables lfun.body in
+    let local_free_vars = Ident.Set.inter free_vars local_idents in
+    let free_vars_block_size, subst, block_fields_rev =
+      Ident.Set.fold (fun var (i, subst, fields) ->
+          let access =
+            Lprim (Pfield (i, Pointer, lifted_block_mut),
+                   [Lvar block_var],
+                   no_loc)
+          in
+          (succ i, Ident.Map.add var access subst, Lvar var :: fields))
+        local_free_vars (0, Ident.Map.empty, [])
+    in
+    let new_fun =
+      lfunction_with_body lfun
+        (Lambda.subst (fun _ _ env -> env) subst lfun.body)
+    in
+    let lifted = { lfun = new_fun; free_vars_block_size } in
```

Comment:
I didn't think it would correspond to silly code (at least, no more than the other recursive definitions we allow). But we wouldn't gain much by making a special case, so it didn't feel worth it.
I'm adding a comment to explain that.